File: fish/colorize_history.rs
use std::process::{Command, Stdio};
use std::io::{self, BufReader, Read, Write};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    const TIME_LEN: usize = 19;
    const PREFIX: &str = "               │ ";
    const BATCH_SIZE: usize = 100;

    let mut todo = true;
    while todo {
        todo = false;
        let mut num_lines = [0u8; BATCH_SIZE];
        let mut times = [[0u8; TIME_LEN]; BATCH_SIZE];

        let mut child = Command::new("fish_indent")
            .arg("--ansi")
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .spawn()?;
        let mut stdin = child.stdin.take().unwrap();
        let stdout = child.stdout.take().unwrap();

        let mut idx = 0;
        let mut c = 0;
        for byte in BufReader::new(io::stdin()).bytes().flatten() {
            if c < TIME_LEN {
                times[idx][c] = byte;
                c += 1;
            } else {
                stdin.write_all(&[if byte == b'\0' { b'\n' } else { byte }])?;
            }
            if byte == b'\n' {
                num_lines[idx] += 1;
            }
            if byte == b'\0' {
                num_lines[idx] += 1;
                idx += 1;
                c = 0;
                if idx >= BATCH_SIZE {
                    todo = true;
                    break;
                }
            }
        }

        drop(stdin);

        let mut color: Vec<u8> = "\x1b[90m".bytes().collect();
        let mut in_esc = false;
        let mut l = 0;
        idx = 0;
        print!("\x1b[90m");
        io::stdout().write_all(&times[0])?;
        for byte in BufReader::new(stdout).bytes().flatten() {
            if byte == b'\n' {
                l += 1;
                if l == num_lines[idx] {
                    print!("\0");
                    idx += 1;
                    l = 0;
                    if idx >= BATCH_SIZE - 1 {
                        break;
                    }
                    print!("\x1b[90m");
                    io::stdout().write_all(&times[idx])?;
                } else if byte == b'\n' {
                    print!("\n\x1b[90m{PREFIX}{}", String::from_utf8_lossy(&color));
                }
            } else {
                io::stdout().write_all(&[byte])?;
                if byte == b'\x1b' {
                    in_esc = true;
                    color.clear();
                }
                if in_esc {
                    color.push(byte);
                    if byte == b'm' {
                        in_esc = false;
                    }
                }
            }
        }
    }
    Ok(())
}

File: fish/colorize_history.py
#!/usr/bin/python3

import os
import re
import secrets
import subprocess
import sys

BATCH_SIZE = 500
TIME_LEN = 19
GRAY = b'\x1b[90m'
PREFIX = GRAY + b'               \xe2\x94\x82 ' # │

DELIM = b'# __HIST_DELIM_' + secrets.token_hex(8).encode() + b'__'

stdout = sys.stdout.buffer
ansi_re = re.compile(rb'\x1b\[[0-9;]*m')


def visible(line):
    return ansi_re.sub(b'', line)


entries = sys.stdin.buffer.read().split(b'\0')
entries.pop() # trailing empty from NUL terminator


def indent_batch(cmds):
    joined = (b'\n\n' + DELIM + b'\n').join(cmds)
    out = subprocess.run(
        ['fish_indent', '--ansi'],
        input=joined,
        capture_output=True,
    ).stdout

    # fish_indent leaves a blank line around the delimiter; skip blanks
    # immediately after each delimiter, then trim trailing blanks per chunk.
    chunks = [[]]
    skip_blank = False
    for line in out.split(b'\n'):
        stripped = visible(line).rstrip()
        if stripped == DELIM:
            chunks.append([])
            skip_blank = True
        elif skip_blank and not stripped:
            continue
        else:
            chunks[-1].append(line)
            skip_blank = False

    for ch in chunks:
        while len(ch) > 1 and not visible(ch[-1]).strip():
            ch.pop()

    return chunks


def emit(time, lines):
    buf = [GRAY, time]
    color = GRAY
    for k, line in enumerate(lines):
        if k > 0:
            buf.extend((b'\n', PREFIX, color))
        buf.append(line)
        m = ansi_re.findall(line)
        if m:
            color = m[-1]
    buf.append(b'\0')

    try:
        stdout.write(b''.join(buf))
    except BrokenPipeError:
        # redirect stdout so interpreter shutdown flush doesn't re-raise
        os.dup2(os.open(os.devnull, os.O_WRONLY), sys.stdout.fileno())
        sys.exit(0)


for i in range(0, len(entries), BATCH_SIZE):
    batch = entries[i : i + BATCH_SIZE]
    times = [e[:TIME_LEN] for e in batch]
    cmds = [e[TIME_LEN:] for e in batch]

    chunks = indent_batch(cmds)

    # Fallback: a malformed entry corrupted the split. Re-indent individually
    # so one bad entry only affects itself.
    if len(chunks) != len(cmds):
        chunks = [indent_batch([cmd])[0] for cmd in cmds]

    for t, ch in zip(times, chunks):
        emit(t, ch)
